04-materials/04-debugging.md

```diff
@@ -1,111 +1,379 @@
 
 ### Instructions
-1. Copy the file at: TBD
-2. Past the file into the file: TBD
-TBD...
 
-## Common Debugging Scenarios
+0. Make sure all your work up to now is saved and/or committed.
+1. Navigate to the `src` directory where your `widget.ts` file is located
+2. Copy the following modified `widget.ts` file contents and paste them into your local `src/widget.ts` file
+
+```{code} typescript
+:linenos:
+:filename: src/widget.ts
+
+import { Widget } from '@lumino/widgets';
+import { MainAreaWidget, ToolbarButton } from '@jupyterlab/apputils';
+import { imageIcon, refreshIcon } from '@jupyterlab/ui-components';
+
+import { requestAPI } from './request';
+
+class ImageCaptionWidget extends Widget {
+  // Initialization
+  constructor() {
+    super();
+
+    // Create and append an HTML <p> (paragraph) tag to our widget's node in
+    // the HTML document
+    const hello = document.createElement('p');
+    hello.innerHTML = 'Hello, world!';
+    this.node.appendChild(hello);
+
+    const center = document.createElement('center');
+    this.node.appendChild(center);
+
+    // Put an <img> tag into the <center> tag, and also save it as a class
+    // attribute so we can update it later.
+    this.img = document.createElement('img');
+    center.appendChild(this.img);
+
+    // Do the same for a caption!
+    this.caption = document.createElement('p');
+    center.appendChild(this.caption);
+
+    // BUG #1: Typo in method name - TypeScript will catch this at build time
+    // Initialize the image from the server extension
+    this.load_imge();
+  }
+
+  // Fetch data from the server extension and save the results to img and
+  // caption class attributes
+  load_image(): void {
+    // BUG #3: Wrong endpoint name - will cause 404 error
+    requestAPI<any>('random-image-captions')
+      .then(data => {
+        console.log(data);
+        this.img.src = `data:image/jpeg;base64, ${data.b64_bytes}`;
+        this.caption.innerHTML = data.caption;
+      })
+      .catch(reason => {
+        console.error(`Error fetching image data.\n${reason}`);
+      });
+  }
+
+  // Information about class attributes for the type checker
+  img: HTMLImageElement;
+  caption: HTMLParagraphElement;
+}
+
+export class ImageCaptionMainAreaWidget extends MainAreaWidget<ImageCaptionWidget> {
+  constructor() {
+    const widget = new ImageCaptionWidget();
+    super({ content: widget });
+
+    this.title.label = 'Random image with caption';
+    this.title.caption = this.title.label;
+    this.title.icon = imageIcon;
+
+    // Add a refresh button to the toolbar
+    const refreshButton = new ToolbarButton({
+      icon: refreshIcon,
+      tooltip: 'Refresh image',
+      onClick: () => {
+        // BUG #2: Wrong method name - using 'as any' bypasses TypeScript checking
+        // This will cause a runtime "is not a function" error when clicked
+        // Should be load_image(), not refresh()
+        (widget as any).refresh();
```

Comment:
Nice idea to show turning off typescript checks can make your life harder! Love this. A good lesson to "avoid `any` at **any** cost" ;)